File: ED del diego/Java/Examenes ED Java/Febrero 2022/Febrero 2022/src/exercises/EdRank.java
/* ------------------------------------------------------------------------------
   -- Student's name:
   -- Student's group:
   -- Identity number (DNI if Spanish/passport if Erasmus):
   --
   -- Data Structures. Grado en Informática. UMA.
   -------------------------------------------------------------------------------
*/

package exercises;

import dataStructures.dictionary.AVLDictionary;
import dataStructures.dictionary.Dictionary;
import dataStructures.graph.DiGraph;
import dataStructures.set.Set;

public class EdRank<T extends Comparable<? super T>> {
    private static final double THRESHOLD = 0.0001;
    private DiGraph<T> diGraph;
    private Dictionary<T, Double> rankDict;

    public EdRank(DiGraph<T> g) {
        diGraph = (DiGraph<T>) g.clone();
        rankDict = new AVLDictionary<>();
        for(T aux : diGraph.vertices()){
            rankDict.insert(aux,0.0);
        }
    }

    public void edRank(double rank) {
        //TODO
        for(T vertex: rankDict.keys()){
            distribute(vertex,rank);
        }
    }

    private void distribute(T node, double rank) {
        //TODO
        if(rank > THRESHOLD){
            double half = rank / (double) 2;
            rankDict.insert(node,rankDict.valueOf(node) + half);
            Set<T> ss = diGraph.successors(node);
            if(ss != null){
                double sRank = half / (double) ss.size();
                for(T vertex: ss){
                    distribute(vertex,sRank);
                }
            }
        }
    }

    public Dictionary<T, Double> edRank() {
        return rankDict;
    }
}

File: ED del diego/Haskell/Examenes/Febrero 2022/haskell/EdRank.hs
-------------------------------------------------------------------------------
-- Student's name:
-- Student's group:
-- Identity number (DNI if Spanish/passport if Erasmus):
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module EdRank where

import DataStructures.Graph.DiGraph
import DataStructures.Dictionary.AVLDictionary as D
import Data.Maybe(fromJust)

threshold = 0.0001
createRankDict :: (Ord a) => DiGraph a -> D.Dictionary a Double
createRankDict g = aux (vertices g) D.empty
    where
        aux [] d = d
        aux (v:vs) d = aux vs (D.insert v 0.0 d)

distribute :: Ord a => DiGraph a -> a -> Double -> D.Dictionary a Double -> D.Dictionary a Double
distribute g v r d = if r <= threshold then d else updateValues g v r' (successors g v) (D.insert v (fromJust (D.valueOf v d) + r') d) -- necesito conocer los sucesores del vértice
    where 
        r' = r / fromIntegral 2

updateValues :: Ord a => DiGraph a -> a -> Double -> [a] -> D.Dictionary a Double -> D.Dictionary a Double
updateValues g v r sucs d = if not (null sucs) then foldl (\ x res -> distribute g res partesIguales x) d sucs else d
    where 
        partesIguales = r / fromIntegral (length sucs)

edRank :: Ord a => DiGraph a -> Double -> D.Dictionary a Double
edRank g r = aux g (vertices g) r (createRankDict g)
    where
        aux g [] r d = d
        aux g (v:vs) r d = aux g vs r (distribute g v r d)